Adds support for parallel inference and batching

TikZSZ opened the pull request:
This PR adds support for running `berkeley-function-call-leaderboard` inference parallelly, `reducing running time by 4x or more` depending on `--batch size`

## Changes
Modifies `berkeley-function-call-leaderboard/model_handler/handler.py`  
- modified `write`  function to make it async using `aiofiles`
- adds `sort_results` function to sort the results based on `idx` after each individual test is over

Modifies `berkeley-function-call-leaderboard/openfunctions_evaluation.py`
- Adds `--batch-size` arg defaults to `1` -> used to controller number of requests made parallelly. 
- Refactors processing and writing logic to `fetch_and_process` function
- Adds `make_async` function to wrap sync function as async -> used to wrap hanlder.inference as async
-  Adds nested progress bar for tracking iterations
- Refactor core logic for processing under `main` function

> This PR takes care of automatically wrapping the inference calls as async, to minimise code changes. The async functions may have side effects on other endpoints currently only openai is tested.

## Testing
Tested on my own custom openai compatible model running on vllm completing simple test in 40 seconds on a RTX4090 LLMA 8B BF16 with batch size 15

ShishirPatil commented:
Thanks for contributing to the Berkeley Function Calling Leaderboard @TikZSZ ! Appreciate the PR and Welcome! We are currently reviewing and testing this PR

TikZSZ commented:
@ShishirPatil I've added proper resume support and updated the PR description with more details about the changes, Could you please take a look when you have a chance? Thank you!
